### PR TITLE
External Storage gametests

### PR DESCRIPTION
**Pull request by starforcraft**
Gametests https://github.com/refinedmods/refinedstorage2/issues/248
External Storage https://github.com/refinedmods/refinedstorage2/issues/75

### CODE REVIEW

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageTest.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+                    new ResourceAmount(asResource(STONE), 2),
+                    new ResourceAmount(asResource(DIRT), 10),
+                    new ResourceAmount(asResource(COBBLESTONE), 3)
+                ))
+                .thenSucceed();
+        });
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldStoreItemAllowlist(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageTest.java]:
```suggestion
    public static void shouldInsertItemAllowlist(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageTest.java [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+                    new ResourceAmount(asResource(STONE), 2),
+                    new ResourceAmount(asResource(WATER), Platform.INSTANCE.getBucketAmount() * 32),
+                    new ResourceAmount(asResource(LAVA), Platform.INSTANCE.getBucketAmount())
+                ))
+                .thenSucceed();
+        });
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldStoreFluidAllowlist(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageTest.java]:
```suggestion
    public static void shouldInsertFluidAllowlist(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageTest.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+            prepareChest(
+                helper,
+                pos.east(),
+                DIRT.getDefaultInstance().copyWithCount(10),
+                COBBLESTONE.getDefaultInstance().copyWithCount(3)
+            );
+
+            externalStorage.setFilterMode(FilterMode.ALLOW);
+            externalStorage.setFilters(Set.of(asResource(STONE)));
+            externalStorage.setPriority(1);
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageTest.java]: Why?

**Review by raoulvdberge — CHANGES_REQUESTED** [review]
I'm still missing a few cases:
- changing of block state should remove from storage (e.g. removing chest block)
- changing of activeness should remove from storage
- extractions through RS seem untested?
- fuzzy mode untested
- priority untested (in combination with other RS storages)
- filter mode -> blocklist?
- access mode untested
- void excess mode
- should detect changes done to the storage itself, without interacting with RS
  - insertion of a new resource
  - addition of an existing resource
  - partial extraction of a resource
  - complete extraction of a resource
  - .. all these types of changes should reflect correctly in RS (see cache impl in `ExternalStorage`)

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageTest.java (hunk 3 above)
Comment (starforcraft) [reply refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageTest.java]: There's an item storage block connected in the network so without setting the priority to 1 the Stone won't be inserted in the chest and will instead go into the item storage block

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java [hunk 4]
```diff
@@ -0,0 +1,745 @@
...
+                    new ResourceAmount(asResource(STONE), 2),
+                    new ResourceAmount(asResource(DIRT), 10),
+                    new ResourceAmount(asResource(COBBLESTONE), 3)
+                ))
+                .thenSucceed();
+        });
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldStoreItemAllowlist(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java]:
```suggestion
    public static void shouldInsertItemAllowlist(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java [hunk 5]
```diff
@@ -0,0 +1,745 @@
...
+                    new ResourceAmount(asResource(DIRT), 12),
+                    new ResourceAmount(asResource(COBBLESTONE), 3),
+                    new ResourceAmount(asResource(STONE), 2)
+                ))
+                .thenSucceed();
+        });
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldStoreItemFuzzyAllowlist(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java]:
```suggestion
    public static void shouldInsertItemFuzzyAllowlist(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java [hunk 6]
```diff
@@ -0,0 +1,745 @@
...
+                    new ResourceAmount(asResource(COBBLESTONE), 3),
+                    new ResourceAmount(asResource(DIAMOND_CHESTPLATE), 1),
+                    new ResourceAmount(asResource(damagedDiamondChestplate), 1)
+                ))
+                .thenSucceed();
+        });
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldStoreItemBlocklist(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java]:
```suggestion
    public static void shouldInsertItemBlocklist(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java [hunk 7]
```diff
@@ -0,0 +1,745 @@
...
+                    new ResourceAmount(asResource(DIRT), 12),
+                    new ResourceAmount(asResource(COBBLESTONE), 3),
+                    new ResourceAmount(asResource(STONE), 2)
+                ))
+                .thenSucceed();
+        });
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldStoreItemFuzzyBlocklist(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java]:
```suggestion
    public static void shouldInsertItemFuzzyBlocklist(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java [hunk 8]
```diff
@@ -0,0 +1,745 @@
...
+                    new ResourceAmount(asResource(STONE), 2),
+                    new ResourceAmount(asResource(DIRT), 5),
+                    new ResourceAmount(asResource(COBBLESTONE), 3)
+                ))
+                .thenSucceed();
+        });
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldRemoveExposedItem(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java]:
```suggestion
    public static void shouldPropagateExternalExtractions(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java [hunk 9]
```diff
@@ -0,0 +1,745 @@
...
+                    pos,
+                    new ResourceAmount(asResource(STONE), 4),
+                    new ResourceAmount(asResource(DIRT), 10)
+                ))
+                .thenSucceed();
+        });
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldPartialRemoveExposedItem(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java]:
```suggestion
    public static void shouldPropagatePartialExternalExtractions(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java [hunk 10]
```diff
@@ -0,0 +1,745 @@
...
+                    new ResourceAmount(asResource(STONE), 4),
+                    new ResourceAmount(asResource(DIRT), 5),
+                    new ResourceAmount(asResource(COBBLESTONE), 3)
+                ))
+                .thenSucceed();
+        });
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldRemoveBrokenExposedItem(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java]:
```suggestion
    public static void shouldNoLongerExposeItemsWhenExternalBlockIsBroken(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java [hunk 11]
```diff
@@ -0,0 +1,745 @@
...
+                    helper,
+                    pos,
+                    new ResourceAmount(asResource(STONE), 2)
+                ))
+                .thenSucceed();
+        });
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldChangedRedstoneModeExposeItem(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java]:
```suggestion
    public static void shouldRespectRedstoneMode(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java [hunk 12]
```diff
@@ -0,0 +1,745 @@
...
+                    pos.east(),
+                    new ResourceAmount(asResource(STONE), 2),
+                    new ResourceAmount(asResource(DIRT), 10),
+                    new ResourceAmount(asResource(COBBLESTONE), 3)
+                ))
+                .thenWaitUntil(storageContainsExactly(
+                    helper,
+                    pos,
+                    new ResourceAmount(asResource(STONE), 2)
+                ))
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java:629]: ok, now turn redstone on and validate (place lever or redstone block or something)

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java [hunk 13]
```diff
@@ -0,0 +1,745 @@
...
+                .thenSucceed();
+        });
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldVoidItem(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java]:
```suggestion
    public static void shouldVoidExcess(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java [hunk 14]
```diff
@@ -0,0 +1,745 @@
...
+                    helper,
+                    pos,
+                    new ResourceAmount(asResource(STONE), 64 * 27)
+                ))
+                .thenSucceed();
+        });
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldNotAllowExtractingItem(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java]:
```suggestion
    public static void shouldRespectInsertAccessMode(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java [hunk 15]
```diff
@@ -0,0 +1,745 @@
...
+                    helper,
+                    pos,
+                    new ResourceAmount(asResource(STONE), 74)
+                ))
+                .thenSucceed();
+        });
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldNotAllowInsertingItem(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java]:
```suggestion
    public static void shouldRespectExtractAccessMode(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java [hunk 16]
```diff
@@ -0,0 +1,745 @@
...
+                    helper,
+                    pos,
+                    new ResourceAmount(asResource(STONE), 4),
+                    new ResourceAmount(asResource(DIRT), 5),
+                    new ResourceAmount(asResource(COBBLESTONE), 3)
+                ))
+                .thenSucceed();
+        });
+    }
+
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java:492]: Missing test: `shouldPropagateExternalInsertions` ?

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java [hunk 17]
```diff
@@ -0,0 +1,745 @@
...
+import static net.minecraft.world.item.Items.STONE;
+
+@GameTestHolder(IdentifierUtil.MOD_ID)
+@PrefixGameTestTemplate(false)
+public final class ExternalStorageItemTest {
+    private ExternalStorageItemTest() {
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldExposeItem(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageItemTest.java]:
```suggestion
    public static void shouldExposeItems(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageFluidTest.java [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+                    new ResourceAmount(asResource(STONE), 2),
+                    new ResourceAmount(asResource(WATER), Platform.INSTANCE.getBucketAmount() * 32),
+                    new ResourceAmount(asResource(LAVA), Platform.INSTANCE.getBucketAmount() * 2)
+                ))
+                .thenSucceed();
+        });
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldStoreFluidBlocklist(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageFluidTest.java]:
```suggestion
    public static void shouldInsertFluidBlocklist(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageFluidTest.java [hunk 19]
```diff
@@ -0,0 +1,194 @@
...
+                    new ResourceAmount(asResource(STONE), 2),
+                    new ResourceAmount(asResource(WATER), Platform.INSTANCE.getBucketAmount() * 32),
+                    new ResourceAmount(asResource(LAVA), Platform.INSTANCE.getBucketAmount())
+                ))
+                .thenSucceed();
+        });
+    }
+
+    @GameTest(template = "empty_15x15")
+    public static void shouldStoreFluidAllowlist(final GameTestHelper helper) {
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageFluidTest.java]:
```suggestion
    public static void shouldInsertFluidAllowlist(final GameTestHelper helper) {
```

File: refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageFluidTest.java [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+                    new ResourceAmount(asResource(LAVA), Platform.INSTANCE.getBucketAmount() * 2)
+                ))
+                .thenWaitUntil(storageContainsExactly(
+                    helper,
+                    pos,
+                    new ResourceAmount(asResource(WATER), Platform.INSTANCE.getBucketAmount() * 32),
+                    new ResourceAmount(asResource(LAVA), Platform.INSTANCE.getBucketAmount() * 2)
+                ))
+                .thenSucceed();
+        });
```
Comment (raoulvdberge) [inline refinedstorage-neoforge/src/test/java/com/refinedmods/refinedstorage/common/storage/externalstorage/ExternalStorageFluidTest.java:194]: What about fluid extractions?

**Review by raoulvdberge — CHANGES_REQUESTED** [review]
Mostly naming ;)

**Review by raoulvdberge — APPROVED** [review]

**Pull request merged**